### DISCUSSION TITLE
Style for Docs and Samples

### DISCUSSION THREAD

**Discussion by christopherswenson** [malloydata/malloy#287] [General]
In https://github.com/looker-open-source/malloy/pull/283 and https://github.com/looker-open-source/malloy/pull/284 I tried to make all our Malloy code have consistent styling/formatting. I think having consistent styling gives our documentation a high level of polish. 

Someday we'll have a linter and auto-formatter, but today is not that day, so for now we should all know the style and abide by it, but in order to do that, we all need to agree on it!

I've written up the rules I've used, along with a bit of justification. Please comment whether you agree or disagree, and we'll discuss.

### Two Space Tab Width

Indentation should be two spaces. Not tabs. Not four spaces.

Example:
```malloy
query: flights -> {
  group_by: carrier
  aggregate: flight_count
}
```

**Justification**: This is what we do elsewhere in the repo. Easy to be consistent.

### Spaces Around Arrows

Arrows should always have spaces around them, whether the query is being referenced by name or being defined.

Examples:
```malloy
query: flights -> flight_count

query: flights -> {
  group_by: carrier
}
```

The one exception is for explores based off other queries, where there is a space after the arrow, but not before.

Example:

```malloy
explore: from(-> some_query) {
  ...
}

query: -> some_other_query -> {
  ...
}
```

**Justification**: I find it easier to read.

### Spaces Around Binary Operators

All binary operators should have spaces around them.

Examples: `x * y`, `x / (1 + 2)`, `floor(x / 2) * 2`, `x > 3000`, `state = 'CA'`, `x = 'CA' | 'NY'`

Exceptions are safe cast (`::`) and Apply (`:`).

**Justification**: I find it easier to read.

### Inline Refinements Should Have Spaces

When a refinement is placed all on one line instead of across multiple lines, there should be a space after the open brace and a space before the close brace.

Example:
```malloy
query: flights { where: distance > 1000 } -> by_carrier
```

**Justification**: I find it easier to read.

### Spaces Before Refinement

There should always be a space before the open brace of refinement.

Examples:
```malloy
explore: flights is table('foo.bar.baz') {
  measure: ca_flight_count is flight_count { where: origin.state = 'CA' }
}
```

**Justification**: I find it easier to read, and I think it looks nicer, and feels familiar coming from C-like languages.

### Subsequent Stages on Same Line

Query stages after the first stage should start on the same line as the close brace of the previous stage.

Example:
```malloy
query: word_score is {
  ...
} -> {
  ...
} -> {
  ...
}
```

It is also okay to have an entire stage on one line, but the next stage should still always start on the same line as the previous ends.

Example:
```malloy
query: foo is bar -> { where: x = 1 } -> {
  aggregate: something
} -> { project: * } -> {
  ...
}
```

**Counterargument**: It's easier to comment out one stage of a pipeline if the arrow is on a new line.
**Justification**: It's easier to read this way. With the arrow on a new line, it makes it look like a new thing at the same level as the query itself. 

### No Leading Commas

Commas should never appear as the first non-whitespace character of a line.

Example:
```malloy
query: foo is flights -> {
  where: [
    thing_one,
    thing_two,
    thing_three
  ]
  ...
}
``` 

Not:
```malloy
query: foo is flights -> {
  where: [
    thing_one
    , thing_two
    , thing_three
  ]
  ...
}
``` 

**Justification**: It looks nicer this way. And since trailing commas are allowed, you can still comment out the last line without causing a compile error.

### Single Line Lists Should Have Surrounding Spaces

Lists that appear all on one line should have a space after the open bracket and before the close bracket.

Example
```malloy
query: foo -> {
  group_by: [ thing_one, thing_two, thing_three ]
}
```

**Justification**: I find it easier to read.

### Single Line Lists With Commas

Lists of things should have commas:

Example: `group_by: [ thing_one, thing_two ]`

Not: `group_by: [ thing_one thing_two ]`

**Justification**: I find it really hard to read without the commas.

### Single Line Lists Should Have Internal Spaces

Lists of things (parameter lists or bracketed lists) should have spaces after the commas:

Example: `nullif(thing, 0)` and `group_by: [ thing_one, thing_two ]`

Not: `nullif(thing,0)` and `group_by: [ thing_one,thing_two ]`

**Justification**: It looks nicer.

### Space After Property Name

There should always be a space after a property name.

Example:
```malloy
query: -> something { where: condition }
```

Not: `query: -> something { where:condition }`

**Justification**: I find it easier to read.

### Only Slash Comments

Use only `//` or `/* */` comments, not `--` comments. 

Example:
```malloy
query: /* foo is */ flights -> {
  // group_by: thing
  group_by: other_thing
}
```

**Justification**: The new Malloy syntax is closer to C-like than SQL-like, so `//` feels more natural.

### Safe Cast Has No Spaces

No spaces around the `::` operator.

Example: `timestamp_function(stuff)::timestamp`

**Justification**: Helps show (high) operator precedence.

### No Space Before Apply Operator

Never put a space before the Apply operator.

Examples: `state: 'CA'`, `thing: pick ...`

**Justification**: It looks weird with a space.

### Lowercase SQL Functions

All SQL functions should be lower-case.

Examples: `nullif(thing, 0)`, `floor(thing)`

**Justification**: All caps SQL functions look outdated.

### No Count Star

Never use `count(*)`. Always use `count()`

**Justification**: It's not necessary, and there's no reason to do it both ways.

### Plural Duration Units

Duration units should be plural unless the quantity is 1.

Examples: `time - 1 year`, `time - 2 months`

**Justification**: `2 year` is grammatically incorrect and looks bad.

### Allowed Semicolon Properties

In multi-line refinements, it is okay to use `;` to put `top`, `order_by`, and `limit` on one line. 

Example:
```malloy
query: foo -> {
  top: 5; order_by: [ size, weight ]
  ...
}
```

Not:
```
query: foo -> {
  group_by: type; aggregate: total_weight
}
```

This does not apply for single-line refinements, e.g. the following is okay:
```malloy
query: foo is -> { group_by: type; aggregate: total_weight }
```

### No Space Before Semicolon

There should not be a space before a semicolon.

Example: `query: foo is -> { group_by: type; aggregate: total_weight }`
Not: `query: foo is -> { group_by: type ; aggregate: total_weight }`

### Separate Filters With Comma When on Different Lines

When multiple filters are used spanning multiple lines, separate with `,`, not `and`. It is okay to put multiple filters together on one line, especially if it makes sense to logically think of them as a unit.

Example:

```
flights {
  where: [
    some_field != null and length(some_field) > 10,
    destination.state = 'CA',
    carrier.nickname = 'Southwest' | 'Delta'
  ]
} -> airport_dashboard
```

Not:
```
flights {
  where: 
    some_field != null and length(some_field) > 10
    and destination.state = 'CA'
    and carrier.nickname = 'Southwest' | 'Delta'
} -> airport_dashboard
```

### When In Doubt: Use Our Typescript Style

This is an incomplete list of rules. There are surely other things that feel obvious to me. If they become inconsistent, we can write them down.

**Comment by anikaks** [comment]
Thanks for getting this discussion started @christopherswenson !! I'm mostly on board with your recommendations, with a few things I'd love to discuss a tad more.

My thoughts/comments:
>The one exception is for explores based off other queries, where there is a space after the arrow, but not before.

I'm not sure I agree with this, if we're doing spaces padding all the other parens punctuation I'd lean toward being consistent here.


>Query stages after the first stage should start on the same line as the close brace of the previous stage.

Let's discuss this one; I agree with you that the alternative looks better, I think it reads similarly clearly, but for me the downside of being less comment-out-able is annoying; my initial reaction is "okay I guess I'll just remove the extra new line before publishing anything" which is probably an iffy sign? But I can get on board if I'm alone feeling this way. I do think that multi-stage operations are quite likely to have individual stages commented out for troubleshooting. 


**Trailing/Leading commas:**
In SQL I am firmly team Leading. However since we allow trailing commas I am willing to change my ways


>Single Line Lists Should Have Surrounding Spaces

This makes me feel uncomfortable for reasons I can't describe. I can get on board but curious if it bothers anyone else.


>Only Slash Comments

I disagree with this one. As a SQL person, the prospect of changing my commenting style for something that still writes SQL is bothersome to me. I have a strong preference for `--` comments. Most data people are not C people, even if Malloy looks like C.


>All caps SQL functions look outdated.

How dare! Actually the main reason I am ok with dropping caps on Malloy functions is that in many cases we are re-working ours and it is a reminder for people to not think of Malloy as being _exactly_ like SQL. My (non-data-backed) opinion is that most SQL people still write caps for functions and keywords when it's for anything they will share or present.

**Reply by christopherswenson** [reply]
**Spaces Around Arrows**
> I'm not sure I agree with this, if we're doing spaces padding all the other parens punctuation I'd lean toward being consistent here.

I just don't like `from( -> some_query)`, because now it looks like there should be a space after `some_query`, too.

**Subsequent Stages on Same Line**
> my initial reaction is "okay I guess I'll just remove the extra new line before publishing anything" which is probably an iffy sign? 

Yeah, I have this reaction, too... It makes me vaguely worried that it's the wrong syntax because it suggests formatting that is inconvenient. I'm still in the camp of: it looks more familiar on the same line, and is more readable, and that trumps the ability to easily comment out, but interested to hear more opinions, too.

**Single Line Lists Should Have Surrounding Spaces**
> This makes me feel uncomfortable for reasons I can't describe.

Ha I get that. I would also be okay with `query: { where: [thing_one, thing_two, thing_three] }`.

**Reply by anikaks** [reply]
>I just don't like from( -> some_query), because now it looks like there should be a space after some_query, too.

I actually see what you're saying here. It slightly bugs me that there's no space in functions and spaces everywhere else but not enough to do anything about it. 


> I'm still in the camp of: it looks more familiar on the same line, and is more readable, and that trumps the ability to easily comment out

My concern with this is that if the "more readable" thing is more annoying to write, people just won't follow the convention. There's a balance for us to find here.

**Reply by christopherswenson** [reply]
@mtoy-googly-moogly made a great point in chat. Our code serves a different purpose than others' code, in that ours is meant to be educational, and therefore readability is higher priority than write/edit-ability.

**Reply by anikaks** [reply]
yes and no - it's meant to be educational and readable, but if we're hoping to set a standard others will want to use it also has to be writeable. I think it's a balance rather than a hard preference in one direction.

**Comment by mtoy-googly-moogly** [comment]
I agree with @christopherswenson  everywhere except upper/lower case function names. I just would not make a rule here.

We currently are just passing all functions the SQL behind us, and in that case it kind of makes sense to UPPERCASE them to emphasize that is is not a thing that malloy knows anything about.

We may someday provide a full expression language, including our own functions, and on that day we would want all function references to be lower.

Meanwhile we shouldn't be opinionated.

**Reply by christopherswenson** [reply]
That's a really interesting point. I can also see an argument for keeping them all lowercase _because_ we ultimately want them to be part of our own expression language, and we want to not draw attention to the fact that they're not yet...

I could also be okay with writing them in all caps. 

But I don't like having no rule here. I think it looks messy and inconsistent.

**Reply by mtoy-googly-moogly** [reply]
if i had to choose between insisting on caps and insisting on lower, i would insist on lower, but I still prefer letting the usr choose for clarity. Ugliness is in the eye of the beholder.

**Reply by christopherswenson** [reply]
From chat, @mtoy-googly-moogly suggests we enforce `UPPER` functions, and then when we have our own, we enforce `lower` for the ones we've implemented. I'm in favor of this idea.

**Reply by anikaks** [reply]
I think I like this, though I'm wondering whether this will actually be quite confusing/appear inconsistent,  because in many cases our functions will be identical to the SQL functions and it won't be clear to folks why some are capitalized and others aren't.

**Comment by mtoy-googly-moogly** [comment]
Proposal ...

Except for putting all `order_by`/`limit`/`top` on one line ... never use `;`

OK:

```
-> {
  top: 5; order_by: [ size, weight ]
  ...
}
```

Not OK:
```
-> {
  group_by: type; aggregate: total_weight
}
```

**Reply by mtoy-googly-moogly** [reply]
actually .. 

```
-> {
  top:5; group_by: size
}
```

Reads pretty well too ... but still .. .some words about `;` are needed

**Reply by christopherswenson** [reply]
I totally agree, except for when the whole refinement is on one line, e.g.
```
query: foo is -> { group_by: type; aggregate: total_weight }
```

**Reply by mtoy-googly-moogly** [reply]
also raises a style question ... is it ..

`top: 5; group_by: ...`

or

`top: 5 ; group_by: ...`

**Reply by christopherswenson** [reply]
Added **No Space Before Semicolon** and **Allowed Semicolon Properties** to discussion above.

Re: **No Space Before Semicolon**

As with the apply operator, I think it looks weird to have a space before `;`, since we don't do that in English.

**Comment by mtoy-googly-moogly** [comment]
@lloydtabb suggests

Pipelines is hard. They way you do it is to comment out a stage, look at the results and go on to the next stage. I understand that it offends typescript aesthetic but there is really good reason to do this.
```
query: word_score is {
  ...
} 
-> {
  ...
} 
-> {
  ...
}
```

**Reply by christopherswenson** [reply]
Our documentation is about teaching people Malloy, so it should be maximally readable. I think that this looks too much like three different things at the top level of the explore, and is therefore harder to read.

```
explore: foo is table(...) {
  query: word_score is {
    ...
  } 
  -> {
    ...
  } 
  -> {
    ...
  }
}
```

**Reply by anikaks** [reply]
I think that `} -> {` is marginally more readable, but substantially harder to deal with when commenting code. So my vote falls with splitting it.

**Reply by christopherswenson** [reply]
But when reading our docs, users don't have to comment out code, but they do have to read and understand it. And if they copy the code into their own file/workspace, they can always add a newline if they want.

**Reply by anikaks** [reply]
i'm thinking of our examples (particularly the /samples models) as being our "standard" of how malloy should look, versus thinking about "what is the most readable documentation" but maybe y'all are thinking of it differently? I just don't think it's _that_ hard to read.

**Reply by christopherswenson** [reply]
Yeah, that's true. Okay, I'm willing to waver on it. Especially since we have syntax highlighting and code lenses to help guide us.

To be clear `x -> y -> { foo } -> { bar }` would still be allowed on one line. It's only when a multi line refinement ends and you want to extend the pipeline, right?

**Reply by anikaks** [reply]
Totally agree!

**Comment by lloydtabb** [comment]
```
explore: flights is table('foo.bar.baz'){
  measure: ca_flight_count is flight_count{ where: origin.state = 'CA' }
}
```

In functions there is no space between the name and the paren.  I think for the same reason we don't want space between the name and the refinement.  It binds tightly to the name.  I'm 75% on this.

**Reply by christopherswenson** [reply]
I think if we want tighter binding, we should use `(` and not `{`. 

I find `flight_count{ where: origin.state = 'CA' }` to look too jammed together, and therefore it's harder to read.

**Reply by christopherswenson** [reply]
Or maybe we add some other syntax to make it easier to see the binding, e.g.

```
flight_count + { where: origin.state = 'CA' }
```

Though that's probably a topic for another day...

**Reply by mtoy-googly-moogly** [reply]
don't want to spin too far off topic, i agree with chris, and i still am looking at the filtering gesture which is currently a refinement to possibly be a totally different thing ... NOT PROPOSING THIS, but just as a for example ..

`ca_flight_count is %{ flight_count ->| origin.state = 'CA' }`

The important concept is a filter ALWAYS encloses the thing it is filtering ... in the interim I am OK with how things are now.

**Reply by anikaks** [reply]
`flight_count{ where: origin.state = 'CA' }` for me, I find this looks funny / unbalanced with a space inside but not before. I could be happy with either

`flight_count { where: origin.state = 'CA' }`
or 
`flight_count{where: origin.state = 'CA'}`

**Reply by anikaks** [reply]
actually I don't like that second option at all

**Reply by lloydtabb** [reply]
I'm ok with spaces

  `ca_count is flight_count { where: origin.state = 'CA' }`

is fine by me.

**Comment by mtoy-googly-moogly** [comment]
We could enforce `foo + 1 year` and `bar + yearCount years` in grammar. currently "year" and "years" are the same symbol

( that is `year` ONLY words when preceded by `1` )

**Reply by christopherswenson** [reply]
I'm not convinced we _need_ to do that. I think just writing `1 year` and `n years` in our code is sufficient.

**Reply by christopherswenson** [reply]
But we certainly could. I guess the argument against is that it's nice to not have to care when you're writing code.

**Reply by anikaks** [reply]
I would be in favor of both `1 year` and `1 years` being valid as far as the language and grammar is concerned because it's annoying to change a filter in 2 places; as a person writing docs/models others will see I will attempt to use the right plural/non-plural of years.

**Comment by lloydtabb** [comment]
We probably need to figure out how we write where blocks with lots and lots of filters.  Do we prefer the example below?  Drilling is going to generate pretty huge where blocks.

```
flights {
  where: [
    origin.code = 'SFO',
    destination.state = 'CA',
    carrier.nickname = 'Southwest' | 'Delta'
  ]
} -> airport_dashboard
```

**Reply by christopherswenson** [reply]
Yes, I think wheres with multiple anded conditions should be written like that. It's okay to logically combine with `and` specifically if it makes sense though, I think, e.g.

```
foo -> {
  where: [
    str != null and length(str) > 5,
    something = 10,
    thingy
  ]
}
```

**Comment by mtoy-googly-moogly** [comment]
FYI: Commas are currently NOT optional between filters in a filter list. I was afraid to make them optional because there is not a keyword like there is in definition lists to disambiguate, might make the parser much slower and the error recovery less effective.

This is a problem that commas are sometimes optional, I know. I was hoping that people would decide that they liked commas everywhere, that didn't happen.

**Comment by christopherswenson** [comment]
@mtoy-googly-moogly Are you okay with arrows on new lines? (any time it would've been `} -> {` it's `}\n  -> {`)

```malloy
explore: foo is table(...) {
  query: word_score is {
    ...
  } 
  -> {
    ...
  } 
  -> {
    ...
  }

  query: another is { group_by: bar } -> { project: * } -> {
    aggregate: baz
  }
  -> { nest: qux }
  -> {
    aggregate: xyzzy
  }
}
```

**Reply by mtoy-googly-moogly** [reply]
I don't like

```Typescript
if (x) {
   ...
}
else {
```

So I also do not like

```
{
  group_by: ...
}
->  {
```

"I comment out sections of code all the time as I am working on my model" is a very common @lloydtabb  technique and I want to listen to that, maybe lloyd really does show us something that is true for data programming which is not true for all other brace based languages.

If you are asking me, "do i like it", the answer is no.

I kind of suspect @lloydtabb might be the only one who wants the `-> {` on a line, but I could be wrong.

**Reply by mtoy-googly-moogly** [reply]
although if we made explore-queries START with `->` then you get ...

```
explore: .... {
  query: word_score is
    -> {
      ...
    } 
    -> {
      ...
    } 
    -> {
      ...
    }
}
```

Which might be better?

**Reply by christopherswenson** [reply]
@anikaks Also likes it. 

I don't like it, but am willing to budge if we can't come to a better answer.

I think the best parallel here is to chaining methods in JS:

```js
someFunction(withSomeParameters)
  .foo(something)
  .bar(
    something,
    somethingElse,
  )
  .baz(thing);
```

Makes me wonder whether `{}` is wrong for queries, and something more like the following would fit our use case better:

```malloy
explore: thing is table('something').refine(
  query: foo is reduce(
    group_by: foo
  )
  .project(
    include: [ bar, baz, qux ]
  )
)
```

Which honestly feels really nice...
